src/hooks/useLinearPrimaryTicks.ts

```diff
@@ -0,0 +1,98 @@
+import type { ScaleContinuousNumeric } from 'd3-scale';
+import { MutableRefObject, useEffect, useMemo, useState } from 'react';
+
+import { textDimentions } from '../utils';
+
+type Directions = 'horizontal' | 'vertical';
+
+export interface PrimaryLinearTicks {
+  label: string;
+  position: number;
+}
+
+interface Options {
+  tickFormat?: (d: number) => string;
+  scientificNotation?: boolean;
+  minSpace?: number;
+}
+
+const TEST_HEIGHT = '+1234567890';
+const MAX_ITERATION = 20;
+
+export function useLinearPrimaryTicks<
+  Scale extends ScaleContinuousNumeric<number, number>
+>(
+  scale: Scale,
+  direction: Directions,
+  ref: MutableRefObject<SVGGElement | null>,
+  options: Options = {},
+): PrimaryLinearTicks[] {
+  const [ticks, setTicks] = useState<number[]>([]);
+
+  const range = scale.range();
+  if (!range) throw new Error('Range needs to be specified');
+
+  const domain = scale.domain();
+  if (!domain) throw new Error('Domain needs to be specified');
+
+  const { scientificNotation = false, minSpace = 4 } = options;
+  const tickFormat: (d: number) => string = useMemo(
+    () =>
+      options.tickFormat || scientificNotation
+        ? (x) => x.toExponential(2)
+        : (x) => JSON.stringify(x),
+    [scientificNotation, options.tickFormat],
+  );
+
+  const axisLength = Math.abs(range[0] - range[1]);
+
+  // Calculates the tick number that fits in the given space
+  useEffect(() => {
+    if (ref.current) {
+      let tickNumber: number | undefined;
+      let repeat = true;
+      let ticks: number[] = [];
+
+      if (direction === 'horizontal') {
+        for (let count = 0; repeat && count < MAX_ITERATION; count++) {
+          // get next ticks
+          ticks = scale.ticks(tickNumber);
+          const formatedTicks = ticks.map(tickFormat);
+          tickNumber = Math.min(ticks.length, tickNumber || Infinity);
```

Comment:
why not use the number of ticks recommanded by d3 if it is higher that the number of ticks passed?

Reply:
This was because it could try to have a "nice" number of ticks that could be bigger, meaning that if I send 12 it will try to plot 13, so when it arrives at the loop, will be looping forever

Reply:
But I think the issue here is that you shouldn't use the number of ticks given by the user, always the number returned by `ticks`. So you should call `ticks` until it returns a new (smaller) value.
Also, what does `d3.ticks` do when you pass it `undefined`?

Reply:
with `undefined` it calculates what is the "best" amount of ticks for the domain/range.

Reply:
I think that I understand your point, this variable keeps track of the current number of valid ticks, you could see this like this example

| tickNumber | ticks.length |
|-|-|
| undefined | 5 |
| 4 | 5 |
| 3 | 3 |
| 2 | 2 |